### ISSUE TITLE
Crash on expression KeyError in Python 3.12

### ISSUE THREAD

**Issue by reagle** [saulpw/visidata#2179]
Sometimes when creating an expression column, I'll mistakenly type a function, thinking it will be applied implicitly to the current column ID. For example, the cursor is in the `name` column and I type `=split(' ')` rather than `=name.split(' ')`. VisiData then crashes, losing my work. My expectation is that VisiData should never crash on a stupid user mistake.

```
Traceback (most recent call last):
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/wrappers.py", line 105, in wrapply
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/column.py", line 340, in getValue
    return self.calcValue(row)
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/column.py", line 556, in calcValue
    r = self.sheet.evalExpr(self.compiledExpr, row, col=self)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/sheets.py", line 361, in evalExpr
    return eval(expr, vd.getGlobals(), contexts)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<expr>", line 1, in <module>
NameError: name 'split' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/sheets.py", line 82, in __getitem__
    i = self.sheet._ordered_colnames.index(colid)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: 0 is not in list

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/sheets.py", line 90, in __getitem__
    c = self._lcm[colid]
        ~~~~~~~~~^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/basesheet.py", line 39, in __getitem__
    return self.locals[k]
           ~~~~~~~~~~~^^^
KeyError: 0

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/basesheet.py", line 289, in checkCursorNoExceptions
    return self.checkCursor()
           ^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/extensible.py", line 79, in wrappedfunc
    r = oldfunc(*args, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/sheets.py", line 602, in checkCursor
    self.calcColLayout()
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/sheets.py", line 627, in calcColLayout
    width = self.calcSingleColLayout(vcolidx, x, minColWidth)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/sheets.py", line 640, in calcSingleColLayout
    col.width = max(col.getMaxWidth(vrows), minColWidth)
                    ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/column.py", line 471, in getMaxWidth
    row_w = dispwidth(self.getDisplayValue(r), maxwidth=self.sheet.windowWidth)
                      ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/column.py", line 422, in getDisplayValue
    return self.getCell(row).text
           ^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/column.py", line 359, in getCell
    cellval = wrapply(self.getValue, row)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/wrappers.py", line 107, in wrapply
    e.stacktrace = stacktrace()
                   ^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/errors.py", line 15, in stacktrace
    return traceback.format_exc().strip().splitlines()
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/traceback.py", line 184, in format_exc
    return "".join(format_exception(sys.exception(), limit=limit, chain=chain))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/traceback.py", line 139, in format_exception
    te = TracebackException(type(value), value, tb, limit=limit, compact=True)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/traceback.py", line 763, in __init__
    suggestion = _compute_suggestion_error(exc_value, exc_traceback, wrong_name)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/traceback.py", line 1087, in _compute_suggestion_error
    list(frame.f_locals)
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/sheets.py", line 96, in __getitem__
    raise KeyError(colid) from e
KeyError: 0

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/sheets.py", line 82, in __getitem__
    i = self.sheet._ordered_colnames.index(colid)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: 0 is not in list

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/sheets.py", line 90, in __getitem__
    c = self._lcm[colid]
        ~~~~~~~~~^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/basesheet.py", line 39, in __getitem__
    return self.locals[k]
           ~~~~~~~~~~~^^^
KeyError: 0

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/reagle/.pyenv/versions/3.12.1/bin/vd", line 6, in <module>
    visidata.main.vd_cli()
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/main.py", line 378, in vd_cli
    rc = main_vd()
         ^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/main.py", line 338, in main_vd
    run(vd.sheets[0])
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/vdobj.py", line 33, in _vdfunc
    return getattr(visidata.vd, func.__name__)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/extensible.py", line 65, in wrappedfunc
    return oldfunc(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/extensible.py", line 65, in wrappedfunc
    return oldfunc(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/mainloop.py", line 307, in run
    ret = vd.mainloop(scr)
          ^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/extensible.py", line 65, in wrappedfunc
    return oldfunc(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/mainloop.py", line 213, in mainloop
    sheet.execCommand(self.keystrokes, keystrokes=self.keystrokes)
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/extensible.py", line 79, in wrappedfunc
    r = oldfunc(*args, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/basesheet.py", line 221, in execCommand
    self.checkCursorNoExceptions()
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/basesheet.py", line 291, in checkCursorNoExceptions
    vd.exceptionCaught(e)
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/errors.py", line 24, in exceptionCaught
    vd.lastErrors.append(stacktrace())
                         ^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/errors.py", line 15, in stacktrace
    return traceback.format_exc().strip().splitlines()
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/traceback.py", line 184, in format_exc
    return "".join(format_exception(sys.exception(), limit=limit, chain=chain))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/traceback.py", line 139, in format_exception
    te = TracebackException(type(value), value, tb, limit=limit, compact=True)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/traceback.py", line 807, in __init__
    context = TracebackException(
              ^^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/traceback.py", line 763, in __init__
    suggestion = _compute_suggestion_error(exc_value, exc_traceback, wrong_name)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/traceback.py", line 1087, in _compute_suggestion_error
    list(frame.f_locals)
  File "/Users/reagle/.pyenv/versions/3.12.1/lib/python3.12/site-packages/visidata/sheets.py", line 96, in __getitem__
    raise KeyError(colid) from e
KeyError: 0
```

**Comment by reagle** [comment]
BTW: when at the expression prompt and I hit tab to autocomplete the column name, it presents them in an odd order that I can't identify. It's not left to right. It'd be nice if it started with current column then it could move: leftward, rightward, or back to 0 then rightward.

**Comment by anjakefala** [comment]
@reagle By "crash", you mean aborts abruptly and returns to the terminal prompt?

I was not able to reproduce that behaviour - I get a column which contains a caught Exception in each cell. We are going to need more contextual and reproduction information.

(VisiData version, environment, and a step-by-step reproduction. If you can make a gif of the behaviour, that would also be helpful.)

**Comment by anjakefala** [comment]
![Peek 2023-12-19 10-47](https://github.com/saulpw/visidata/assets/7489659/0884472d-9d2c-4dac-a0a1-a1b34715e8be)

**Comment by reagle** [comment]
Hi @anjakefala. I crash to shell prompt. I just updated from github:

|   Resolved https://github.com/saulpw/visidata.git to commit fde05882e205a902b5890212389571b47c61aae4
| macOS 14.2 23C64 arm64
| python 3.12.1 via pyenv 



https://github.com/saulpw/visidata/assets/596814/6c0868c1-9a65-4c23-b4d1-5f668f4e328a

**Comment by anjakefala** [comment]
@reagle Thanks! I still can't reproduce.

Does this happen if you run `vd -N`? (That will turn off all plugins, and code in your `~/.visidatarc`.)

**Comment by reagle** [comment]
Yes @anjakefala, still happens with `-N`.

**Comment by reagle** [comment]
@anjakefala, btw, did you try with a pyenv install or with your typical conda install?

**Comment by reagle** [comment]
I am not encountering this issue on 3.11.6.

**Comment by anjakefala** [comment]
Wonderful!

**Comment by anjakefala** [comment]
So we need to figure out what is causing this with Python 3.12. I'll see if I can reproduce!

**Comment by reagle** [comment]
I can't imagine how it'd be connected to [ncurses widechar support](https://github.com/python/cpython/issues/111460), but the "works on 3.11 but not 3.12" symptom is similar.

**Comment by anjakefala** [comment]
Good news: I reproduced with Python 3.12.

I still don't understand what is going on, but I'm on the case!

**Comment by anjakefala** [comment]
Thank you so much for your patience @reagle! The fix is going to make it into VisiData 3.0.1. `traceback.py` added a `compute_suggestion_error` that expected our frame locals to be iterable.

**Comment by reagle** [comment]
Confirmed! Thank you.
